### PR TITLE
[Feature] Add an option to search input files recursively in ingestion job. The default is set to true to be backward compatible.

### PR DESCRIPTION
**Pull request by 61yao**
searchRecursively option is added in jobSpec. It is added to prevent the non-necessary recursive search for input files. 
Currently, the option is set to true by default since that is existing behavior. 
In next PR, we will change the default to be false after clients are aware of this change.

### CODE REVIEW

**Comment by yaostartree** [comment]
I don't think it is safe to add this flag and set the default option to false since current behavior is searching recursively by default. It is dangerous to change default behavior since we don't know what we will break.

**Comment by yaostartree** [comment]
Updated the default searchRecursive option to be true.

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java [hunk 1]
```diff
@@ -229,4 +233,63 @@ private static String fetchUrl(URL url, String authToken)
...
+   * @param fileUri root directly uri
+   * @param includePattern optional glob patterns for files to include
+   * @param excludePattern optional glob patterns for files to exclude
+   * @param searchRecrusively if ture, search files recursively from directory specified in fileUri
+   * @return list of matching files.
+   * @throws IOException on IO failure for list files in root directory.
+   * @throws URISyntaxException for matching file URIs
+   * @throws RuntimeException if there is no matching file.
+   */
+  public static List<String> listMatchedFilesWithRecursiveOption(PinotFS pinotFs, URI fileUri, String includePattern,
```
Comment (walterddr) [inline pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java:254]: nice refactoring. this avoids duplicate code in many ingestion job runners

File: pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java [hunk 2]
```diff
@@ -56,6 +57,8 @@ public void testIngestionJobLauncherWithUnicodeCharForMultivalueFieldDelimiter()
         GroovyTemplateUtils.class.getClassLoader().getResource("ingestion_job_spec_unicode.yaml").getFile(), null,
         null, null);
     Assert.assertEquals("\ufff0", spec.getRecordReaderSpec().getConfigs().get("multiValueDelimiter"));
+    // searchRecursively is set to false by default.
+    Assert.assertTrue(spec.isSearchRecursively());
```
Comment (walterddr) [inline pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java:61]: the comment doesn't match code

File: pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java [hunk 3]
```diff
@@ -48,6 +48,7 @@ public void testIngestionJobLauncherWithTemplate() {
         null, context, _defaultEnvironmentValues);
     Assert.assertEquals(spec.getInputDirURI(), "file:///path/to/input/2020/05/06");
     Assert.assertEquals(spec.getOutputDirURI(), "file:///path/to/output/2020/05/06");
+    Assert.assertFalse(spec.isSearchRecursively());
```
Comment (walterddr) [inline pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java:51]: make a comment explicitly saying this is set to false in yaml config explicitly

**Review by walterddr — COMMENTED** [review]
thanks for the contribution @yaostartree , this looks great overall. I have some minor comments. please kindly take a look

File: pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java (hunk 3 above)
Comment (yao61) [reply pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java:52]: Updated the comment

File: pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java (hunk 2 above)
Comment (yao61) [reply pinot-spi/src/test/java/org/apache/pinot/spi/ingestion/batch/IngestionJobLauncherTest.java]: Updated

**Review by walterddr — APPROVED** [review]
lgtm.

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java [hunk 4]
```diff
@@ -143,13 +141,6 @@ public void run()
       PinotFSFactory.register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(), new PinotConfiguration(pinotFSSpec));
     }
 
-    //Get pinotFS for input
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:146]: Let's not move this. We want to throw exception before creating output dir when the input is not valid. We may move the list files logic here

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java [hunk 5]
```diff
@@ -45,6 +45,11 @@ public class SegmentGenerationJobSpec implements Serializable {
    */
   private String _inputDirURI;
 
+  /**
+   * If true, search input files recursively from root directly specified in _inputDirURI.
+   */
+  // TODO: set the default value to false after all clients are aware of this.
+  private boolean _searchRecursively = true;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java:52]: (nit) Add an empty line after this

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java [hunk 6]
```diff
@@ -161,6 +166,13 @@ public void setInputDirURI(String inputDirURI) {
     _inputDirURI = inputDirURI;
   }
 
+  public boolean isSearchRecursively() {
+    return _searchRecursively;
+  }
+
+  public void setSearchRecursively(boolean searchRecursively) {
+    _searchRecursively = searchRecursively;
+  }
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java:175]: (nit) Add an empty line

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java [hunk 7]
```diff
@@ -229,4 +233,63 @@ private static String fetchUrl(URL url, String authToken)
     }
     return IOUtils.toString(connection.getInputStream(), StandardCharsets.UTF_8);
   }
+
+
+  /**
+   * Find matching files from root directory specified in fileUri.
+   * If includePattern and excludePattern are not null, get all the files that match includePattern and exclude files
+   * that match excludePattern.
+   * If
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java:242]: (nit) Remove

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java [hunk 8]
```diff
@@ -229,4 +233,63 @@ private static String fetchUrl(URL url, String authToken)
...
+   *
+   * @param pinotFs root directly fs
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java:244]: (minor) You mean `directory` here? Same for `fileUri`

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-2.4/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark/SparkSegmentGenerationJobRunner.java [hunk 9]
```diff
@@ -128,14 +126,6 @@ public void run()
     for (PinotFSSpec pinotFSSpec : pinotFSSpecs) {
       PinotFSFactory.register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(), new PinotConfiguration(pinotFSSpec));
     }
-
-    //Get pinotFS for input
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-2.4/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark/SparkSegmentGenerationJobRunner.java:132]: Same here. Let's first gather the input files before processing the outputDir

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3.2/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentGenerationJobRunner.java [hunk 10]
```diff
@@ -128,14 +126,6 @@ public void run()
     for (PinotFSSpec pinotFSSpec : pinotFSSpecs) {
       PinotFSFactory.register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(), new PinotConfiguration(pinotFSSpec));
     }
-
-    //Get pinotFS for input
```
Comment (Jackie-Jiang) [inline pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3.2/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentGenerationJobRunner.java:132]: Same here

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java (hunk 1 above)
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java:254]: Annotate `includePattern` and `excludePattern` as `Nullable`

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java [hunk 11]
```diff
@@ -32,6 +32,7 @@ public class BatchConfig {
 
   private final FileFormat _inputFormat;
   private final String _inputDirURI;
+  private final boolean _searchRecursively;
```
Comment (Jackie-Jiang) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java:35]: Is this used? If so, the default for it should also be `true`

**Review by Jackie-Jiang — COMMENTED** [review]
Good job extracting the common code into the util

Is the `pinot-plugins.tar.gz` file required for this PR? I feel it is mistakenly included

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java (hunk 4 above)
Comment (61yao) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-hadoop/src/main/java/org/apache/pinot/plugin/ingestion/batch/hadoop/HadoopSegmentGenerationJobRunner.java:146]: Updated

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java (hunk 5 above)
Comment (61yao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java:52]: Done

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java (hunk 6 above)
Comment (61yao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/spec/SegmentGenerationJobSpec.java:176]: Done

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java (hunk 7 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java]: Done

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java (hunk 8 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java]: Yeah. Typo

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-2.4/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark/SparkSegmentGenerationJobRunner.java (hunk 9 above)
Comment (61yao) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-2.4/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark/SparkSegmentGenerationJobRunner.java:132]: Done

File: pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3.2/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentGenerationJobRunner.java (hunk 10 above)
Comment (61yao) [reply pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3.2/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentGenerationJobRunner.java:132]: Done

File: pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java (hunk 1 above)
Comment (61yao) [reply pinot-common/src/main/java/org/apache/pinot/common/segment/generation/SegmentGenerationUtils.java]: Done

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java (hunk 11 above)
Comment (61yao) [reply pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java]: Done

**Comment by 61yao** [comment]
> 

Yeah. Removed it.

**Review by Jackie-Jiang — APPROVED** [review]
LGTM, great job!

**Pull request merged**